server/src/graphql/mutations/new_transaction.rs

```diff
@@ -0,0 +1,170 @@
+    validators::{InputValueValidator, IntNonZero, ListMinLength},
+    FieldError, FieldResult, Value, ID,
+};
+use diesel::prelude::*;
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use tokio_diesel::*;
+use uuid::Uuid;
+
+struct ChangesSumToZero {}
+
+impl InputValueValidator for ChangesSumToZero {
+    fn is_valid(&self, value: &Value) -> Option<String> {
+        if let Value::List(v) = value {
+            let mut sum = 0;
+            for detail in v {
+                if let Value::Object(o) = detail {
+                    if let Some(Value::Int(n)) = o.get("changeCents") {
+                        sum += n
+                    }
+                }
+            }
+
+            if sum != 0 {
+                return Some(String::from("Changes must add up to zero"));
+            }
+        }
+
+        None
+    }
+}
+
+#[async_graphql::InputObject]
+pub struct BalanceChangeDetail {
+    pub balance_id: ID,
+    #[field(validator(IntNonZero))]
+    pub change_cents: i32,
+}
+
+pub struct ParsedBalanceChangeDetail {
+    pub balance_uid: Uuid,
+    pub change_cents: i32,
+}
+
+impl TryFrom<BalanceChangeDetail> for ParsedBalanceChangeDetail {
+    type Error = FieldError;
+
+    fn try_from(value: BalanceChangeDetail) -> FieldResult<ParsedBalanceChangeDetail> {
+        let balance_uid =
+            Uuid::parse_str(&value.balance_id).or_else(|_e| Err(FieldError::from("Invalid ID")))?;
+
+        Ok(ParsedBalanceChangeDetail {
+            balance_uid,
+            change_cents: value.change_cents,
+        })
+    }
+}
+
+#[async_graphql::InputObject]
+pub struct NewTransactionInput {
+    pub squad_id: ID,
+    #[field(validator(and(ListMinLength(length = "1"), ChangesSumToZero)))]
+    pub balance_changes_detail: Vec<BalanceChangeDetail>,
+}
+
+pub struct ParsedNewTransactionInput {
+    pub squad_uid: Uuid,
+    pub balance_changes_detail: HashMap<Uuid, i32>,
+}
+
+impl TryFrom<NewTransactionInput> for ParsedNewTransactionInput {
+    type Error = FieldError;
+
+    fn try_from(value: NewTransactionInput) -> FieldResult<ParsedNewTransactionInput> {
+        let squad_uid =
+            Uuid::parse_str(&value.squad_id).or_else(|_e| Err(FieldError::from("Invalid ID")))?;
+
+        Ok(ParsedNewTransactionInput {
+            squad_uid,
+            balance_changes_detail: value
+                .balance_changes_detail
+                .into_iter()
+                .map(|b| {
+                    let parsed = ParsedBalanceChangeDetail::try_from(b)?;
+
+                    Ok((parsed.balance_uid, parsed.change_cents))
+                })
+                .collect::<FieldResult<HashMap<Uuid, i32>>>()?,
+        })
+    }
+}
+
+#[async_graphql::SimpleObject]
+pub struct NewTransactionPayload {
+    pub squad: Squad,
+    pub transaction: Transaction,
+}
+
+pub async fn new_transaction(
+    pool: &Pool,
+    input: ParsedNewTransactionInput,
+) -> Result<NewTransactionPayload> {
+    Ok(pool
+        .transaction(move |conn| {
```

Comment:
nit: Can we break it up into smaller pieces

`addNode();
 addTransaction();
 updateBalances();`

Feel free to ignore this

Reply:
I'd like to break this up for sure, but I don't know yet what the appropriate level of granularity is - it's tricky because we need some of these operations to be grouped together both logically (we never want to add a node without actually adding the detail for that node) and SQL-transaction-ally.

I'll open an issue to revisit this